src/main/java/com/cloudbees/jenkins/support/filter/FilteredOutputStream.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.jenkins.support.filter;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CoderResult;
+import java.nio.charset.CodingErrorAction;
+import java.nio.charset.StandardCharsets;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Wraps an OutputStream by filtering written lines using a provided ContentFilter.
+ * Data written to an instance of this will be decoded on the fly using the provided charset, and each line
+ * written is filtered.
+ *
+ * @author Matt Sicker
+ * @see ContentFilter
+ * @since 2.48
+ */
+public class FilteredOutputStream extends FilterOutputStream {
```

Comment:
On one hand, I think this class has too much complexity for something "simple" but OTOH, I didn't find any existing class doing that and do not have a better approach.

Perhaps separate into multiple classes, one that convert bytes to characters and the second that takes characters and separate them by line and finally the one that filter whole line using the ContentFilter. (Or perhaps just merge the two last). Just an idea to reduce the complexity by using composition, but not sure it's 100% applicable here.

Reply:
I think what I initially tried was essentially implementing this as a `Writer`, then using a chain like `new WriterOutputStream(new AnonymizedWriter(new OutputStreamWriter(out)))` which started to get silly.

Reply:
If you can do that within 5-10 lines it's a lot clearer than implementing the things at the byte level from my PoV. Just don't loose hours on that :)

Reply:
Playing with charset encoders/decoders and byte buffers is like the majority of the work log4j spends its CPU time doing. Too much practice. ;)